feat(assembly): Refactor `Assembler` to support incremental compilation

frisitano opened the pull request:
This is a draft implementation of the `materialize_module(..)` procedure for the `Assembler`.  This attempts to address the requirements of #894.

Introduce `materialize_module(..)` method in `Assembler`.

bobbinth commented:
Not a review yet, just some initial thoughts about this.

I think there are 4 main use cases for why we need something like this:
1. We need to compute account code roots, and thus, we need to compile all of account methods.
2. We need to compute a script root for a given note.
3. We need to build an executable program for a given note.
4. We may want to check if a given note can be executed against a given account.

For the first point, we can use changes in this PR (though, I wonder if some other approach might be simpler).

The second and third points are more complex though, and require a bit more analysis. There are several cases to consider here, but in all of them we start out with a note script AST and need to get program MAST.

The AST could two types of external references:
- `procedure_id` which is a hash of an absolute path for a given procedure.
- `mast_root` this is what is described [here](https://github.com/0xPolygonMiden/miden-base/discussions/7#discussioncomment-4489315) (not yet implemented).

Let's consider procedure IDs, first. Procedures could come from two places:
- Some well-known library (e.g., `miden-lib`).
- Some account-specific module (i.e., a unique procedure defined in some account).

For the first case, things should be fairly straight-forward: we just need to load the library into the assembler (e.g., using `with_library()` method). Then, calling `compile()` against note script MASM should work, though we should probably expose something like `compile_ast()` as we won't have source MASM code for serialized notes. So, in this case, we don't actually need to have a materialized account module.

For the second case, we don't have a solution right now as we don't have account-specific procedure IDs. But I'm actually wondering if having account-specific procedure IDs is a good idea. A better solution may be to make calls to account-specific methods via `mast_root`s.

There are two was to handle `mast root` approach:
1. We load account modules into the assembler directly (basically, maintain a map similar to `ProcedureCache` in the assembler but for `mast root |-> code block` mapping). This could be done by adding something like `load_module(module_ast)` function to the assembler.
2. We build something like materialized module before hand, and then have something like `compile_with_module(program_ast, mateiralized_module)` method on the assembler.

My (very) preliminary thinking is that the first approach is better. If so, we may be able to get a way with adding just `load_module()` method which would look as follows:

```Rust
pub fn load_module(&mut self, ast: ModuleAst) -> Result<Vec<Digest>, AssemblyError>
```
This method would compile procedures in the provided module, add them to this new cache structure I mentioned above, and return MAST roots of all procedures exported from the module.

Once a module is loaded into the assembler, we would be able to compile note scripts against it by simply calling `compile()` (or `compile_ast()`) method.

bobbinth commented:
A few more thoughts on this.

### 1. Building account code root
The starting point for building account code root is a MASM file which defines account interface. For example, the file could be `my_account.masm` and it could look like so:
```
export.receive_assets
  ...
end

export.send_assets
  ...
end

export.auth_tx
  ...
end
```

If `receive_assets` and `send_assets` are from a standardized interface, then ideally, this would look something like this:
```
use.miden::abi::basic_wallet

export.basic_wallet::receive_assets
export.basic_wallet::send_assets

export.auth_tx
   ...
end
```

Once we have the file defining account interface, we need to convert it to an AST representation. This can be done via [ModuleAst::parse()](https://github.com/0xPolygonMiden/miden-vm/blob/next/assembly/src/parsers/mod.rs#L184) function. This will return a `ModuleAst` struct. This struct has a couple of purposes:

1. Serialized version of this struct would go on chain for public accounts.
2. It would be stored as part of the account code (e.g., [here](https://github.com/0xPolygonMiden/miden-base/blob/main/objects/src/accounts/code.rs#L17)) locally on client devices (e.g., by wallets).
3. It would be used to compute account code root.

For this last point, we'd probably need to add `load_module()` method for the assembler I mentioned in the previous comment. This method would return a vector of MAST roots for all procedures exported by the parsed module AST. We could then sort these roots and build a Merkle tree out of them. The root of this tree would be the account code root.

One open question is which libraries should the assembler be initialized with. We should probably include `stdlib` and `miden-lib`, but maybe something beyond that would be useful as well.

### 2. Building note script root
The starting point for building note script root is also a MASM file. This MASM file will contain an executable program which will call some of account methods. For example, it could look something like this:
```
use.miden::abi::basic_wallet

begin
  ...
  call.basic_wallet::receive_assets
  ...
end
```
The above code could be executed against any account which exposes `basic_wallet::receive_assets` endpoint.

One open question is how scripts which need to call non-standardized account endpoints. One way to support this is to introduce a variant of `call` instruction which would work with raw MAST roots. This could look like so:
```
begin
  ...
  call.0x1234567
  ...
end
```
Where `0x1234567` would be a MAST root of a procedure exposed by the account interface. Note, that we could call standardized procedures in the same way too as all of them would have a specific MAST root (e.g., `basic_wallet::receive_assets` would reduce to some specific MAST root).

Once we have the MASM file, the first thing we need to do is to build an AST out of it. This could be done via the [ProgramAst::parse()](https://github.com/0xPolygonMiden/miden-vm/blob/next/assembly/src/parsers/mod.rs#L59) function. This function returns `ProgramAst` struct. This struct, similarly to `ModuleAst` has a couple of purposes:

1. Serialized version of this struct would go on chain for public notes.
2. It would be stored as part of the note script (e.g., [here](https://github.com/0xPolygonMiden/miden-base/blob/main/objects/src/notes/script.rs#L7)) locally on client devices (e.g., by wallets).
3. It would be used to compute note script root.

For the last point, we'd need to add a version of [compile()](https://github.com/0xPolygonMiden/miden-vm/blob/next/assembly/src/assembler/mod.rs#L124) method to the assembler to accept a `ProgramAst` struct rather than raw source string. Once the `ProgramAst` is compiled into a `Program` struct, we would be able to get the code root from it.

Note that if the note calls any non-standardized account methods, then account module would have be loaded into the assembler before we try to compile note script.

### 3. Building transaction program
To execute a transaction we first need to build a transaction program from note scripts and an optional tx script. Putting aside tx script for now (handling of which would be pretty similar to note scripts anyway), there are two was we can go about building a transaction program:

1. We could build individual script `Program` objects and then merge them together. This would probably be the simplest approach (since it would requite minimal changes in the assembler), but it would entail some inefficiencies. The main inefficiency would come from duplicate data in the `cb_table` contained in individual note programs.
2. We could modify the assembler to enable "incremental" compilation of multiple programs. This could be done by letting the user control the `AssemblyContext`. For this, we could add something like `compile_with_context()` method to the assembler which could look like so:

```Rust
pub fn compile_with_context(
    &self,
    program: &ProgramAst,
    context: &mut AssemblyContext
) -> Result<CodeBlock, AssemblyError>
```
This method would compile the provided program in the specified context and return the MAST (represented by `CodeBlock`) of the program. But it would not reduce the context to `cb_table`. This way, the same context could be used to compile multiple programs, and in the end reduce the context to `cb_table` manually via [AssemblyContext::into_cb_table()](https://github.com/0xPolygonMiden/miden-vm/blob/next/assembly/src/assembler/context.rs#L225).

frisitano commented:
> One open question is which libraries should the assembler be initialized with. We should probably include `stdlib` and `miden-lib`, but maybe something beyond that would be useful as well.

We have discussed the potential of allowing the use of shared user defined libraries, I think this is an interesting idea.  A library should be uniquely defined by a root.  We can use this root as a key for libraries in a (root -> Library) map.  Maybe we can have a `use` declaration that allows the user to specify a root which identifies a shared library.  It would then be the responsibility of the operator to load this library into the assembler for compilation. 

> Note that if the note calls any non-standardized account methods, then account module would have be loaded into the assembler before we try to compile note script.

Could you elaborate a bit on this please?  What do you mean exactly when you refer to standard and non-standard account methods?   Another point that may be relevant to this conversation is private accounts where the account methods are private.  In this case the sender may only know the MAST root of the desired procedure of the receiver account code which has been provided to them out of band by the receiver. In this case we will not be able to load the account procedure into the assembler when compiling the note program.  I guess it would be the responsibility of the receiver to do this when they want to consume the note?

bobbinth commented:
> We have discussed the potential of allowing the use of shared user defined libraries, I think this is an interesting idea. A library should be uniquely defined by a root. We can use this root as a key for libraries in a (root -> Library) map. Maybe we can have a `use` declaration that allows the user to specify a root which identifies a shared library. It would then be the responsibility of the operator to load this library into the assembler for compilation.

Yep - the overall idea makes sense, though there are still some specific open questions in my mind. For example, where definition of such libraries live (i.e., on chain or off chain)? Who can update/create new libraries and how? How a particular note/account can specify upfront (i.e., without requiring parsing) which libraries it needs?

> What do you mean exactly when you refer to standard and non-standard account methods?

By standardized methods I mean methods from interfaces which are well-know. For example, `miden::abi::basic_wallet` might be one such interface. It is well-known because it comes from `miden-lib` so, probably everyone will have it. I could imagine that in the future we might have something like an improvement proposal mechanism (e.g., MIP or PIP) to standardize more interfaces. These could go into `miden-lib` or maybe even some different library.

> Another point that may be relevant to this conversation is private accounts where the account methods are private. In this case the sender may only know the MAST root of the desired procedure of the receiver account code which has been provided to them out of band by the receiver. In this case we will not be able to load the account procedure into the assembler when compiling the note program.

Good point. We should probably modify the assembler to not error out in case some entries in the `cb_table` are missing. These entries may be required for executing the program, but they may not be needed for computing the MAST root of the program. An example of this would be `call.0x1234567` instruction where we have all the required info to compute MAST root in the instruction itself, but we won't be able to execute this instruction unless we know the program under `0x1234567` root.

> I guess it would be the responsibility of the receiver to do this when they want to consume the note?

Yes, the receiver would need provide the implementation for any program which needs to be executed. One thing to note here is that we only need to know the path within the MAST which will be executed - we don't need to know the full program. This is not yet supported by the implementation (as we can't have "hidden" branches in the MAST just yet), but this is one of the reasons for something like [Proxy](https://github.com/0xPolygonMiden/miden-vm/blob/next/core/src/program/blocks/proxy_block.rs) block (which is currently unused).

bobbinth reviewed (commented):
Thank you! I left a couple of comments inline. One other thing I was thinking (though probably for a different PR): we should probably change `ProcedureCache` into a real struct. One reason for this is that we'd want to be able to look up procedures by two things: procedure ID and procedure MAST root.

The interface could look something like:
```Rust
impl ProcedureCache {

    pub fn insert(&mut self, proc: Procedure) {
        todo!()        
    }

    pub fn get_by_id(&self, id: ProcedureId) -> Option<&Procedure> {
        todo!()
    }

    pub fn get_by_hash(&self, root: RpoDigest) -> Option<&Procedure> {
        todo!()
    }

   pub fn contains_id(&self, id: ProcedureId) -> bool {
        todo!()
    }

    pub fn contains_hash(&self, root: RpoDigest) -> bool {
        todo!()
    }
}
```
This let us support `call.<MAST root>` instruction - thought, we'd probably also need to update how `Callset` works as well.

frisitano commented:
I've updated the `Assembler` to support incremental compilation that should satisfy our requirements for building the transaction program and associated assets.

## 1. Building account code root

The account code module can be built using the [`compile_module(..)`](https://github.com/0xPolygonMiden/miden-vm/blob/frisitano-procedure-access/assembly/src/assembler/mod.rs#L193) method.  This method returns `Vec<RpoDigest>` which are the roots of exported procedures.  We can then use this vector to construct the method tree and root for the `AccountCode`.  At the end of this method the account module methods will be added to the assembler procedure cache for downstream usage in note scripts.

## 2. Building note script root

To support the compilation of note scripts we have introduced the [`compile_in_context(..)`](https://github.com/0xPolygonMiden/miden-vm/blob/frisitano-procedure-access/assembly/src/assembler/mod.rs#L151)  method.  This method takes a `ProgramAst`, compiles it and returns a `CodeBlock` which represents the root of the body of the note script.  We want to make the note script body available downstream such that the transaction program can call the note scripts.  To achieve this when we compile the program body we add it to the procedure cache.  The prior `AssemblyContext` enforced that all programs would have a path of `#exec` and the program body would have a procedure name of `#main`. This results in some challenges as we have many notes each containing their own body that we want to compile without naming conflict.  As such we have introduced a constructor to the `AssemblyContext` - [`AssemblyContext::new_program(..)`](https://github.com/0xPolygonMiden/miden-vm/blob/frisitano-procedure-access/assembly/src/assembler/context.rs#L45) such that we can specify the procedure name and path to avoid conflicts when many executable programs incrementally.

## 3. Building transaction program
The requirements here are to ensure that the `Assembler` has the appropriate procedures in it's procedure cache (including the note script bodies) - this is achieved by incrementally compiling the note scripts without consuming the `Assembler`.  

A test case has been provided that shows the expected compilation pipeline for transaction program and associated assets.

frisitano commented:
> Thank you! I left a couple of comments inline. One other thing I was thinking (though probably for a different PR): we should probably change `ProcedureCache` into a real struct. One reason for this is that we'd want to be able to look up procedures by two things: procedure ID and procedure MAST root.
> 
> The interface could look something like:
> 
> ```rust
> impl ProcedureCache {
> 
>     pub fn insert(&mut self, proc: Procedure) {
>         todo!()        
>     }
> 
>     pub fn get_by_id(&self, id: ProcedureId) -> Option<&Procedure> {
>         todo!()
>     }
> 
>     pub fn get_by_hash(&self, root: RpoDigest) -> Option<&Procedure> {
>         todo!()
>     }
> 
>    pub fn contains_id(&self, id: ProcedureId) -> bool {
>         todo!()
>     }
> 
>     pub fn contains_hash(&self, root: RpoDigest) -> bool {
>         todo!()
>     }
> }
> ```
> 
> This let us support `call.<MAST root>` instruction - thought, we'd probably also need to update how `Callset` works as well.

This makes sense to me.  Looking to the future we will also need to be mindful of being able to support the `pcall` operation.  The parser will not be able to infer the `pcall` target from the AST so it will be the responsibility of the client to specify the expected `pcall` targets to ensure they are included in the `Callset`.

bobbinth commented on assembly/src/tests.rs:
```diff
@@ -105,6 +108,71 @@ fn span_and_simple_if() {
+    let _note_2_root = assembler.compile_in_context(note_2, &mut assembler_context).unwrap();
+
+    // compile transaction program
+    let source = "\
+    use.context::note1 
+    use.context::note2
+    
+    # these calls will need to be modified to use MAST roots
+    begin call.note1::main call.note2::main end";
```
I'm actually not sure this is the right way to build transaction programs. Two reasons for this:

First, transaction programs need to include prologue, epilogue, note setup script etc. So, the way this should look like would be:
```
use.miden::sat::internal::prologue
use.miden::sat::internal::epilogue
use.miden::sat::internal::note_setup
use.context::note1 
use.context::note2

begin
  exec.prologue::prepare_transaction

  exec.note_setup::prepare_note
  call.note1::main
  
  exec.note_setup::prepare_note
  call.note2::main

  exec.epilogue::finalize_transaction
end
```

Second, I'm not sure we actually ever need to deal with textual representation of the above. The way I'm imagining we'd build a transaction program is not by writing the above MASM and compiling it, but by combining blocks manually. Specifically:

- We'd have a pre-build `CodeBlock`'s for prologue, epilogue, and note setup script (together with their associated callsets).
- For each transaction, we'd build `CodeBlock`'s for each note using `compile_in_context()`.
- And then we'd "manually" join these blocks together using constructors of `new_join()` and `new_call()` constructors of `CodeBlock` (we'd also need to merge their respective callsets).

To perform the above procedure, we could have a component in `miden-base` (or somewhere else) - maybe something like `TxCompiler` which would internally maintain pre-built `CodeBlock`'s for prologue, epilogue etc. as well as an instance of the assembler.

Btw, with the introduction of `pcall` much of the above complexity would go away as we'd have a single program for transaction execution which could look something like:
```
use.miden::sat::internal::prologue
use.miden::sat::internal::epilogue
use.miden::sat::internal::note_setup

begin
  exec.prologue::prepare_transaction

  while.true
    exec.note_setup::prepare_note
    pcall
  end

  exec.epilogue::finalize_transaction
end
```

frisitano replied:
> First, transaction programs need to include prologue, epilogue, note setup script etc. So, the way this should look like would be:
> 
> ```
> use.miden::sat::internal::prologue
> use.miden::sat::internal::epilogue
> use.miden::sat::internal::note_setup
> use.context::note1 
> use.context::note2
> 
> begin
>   exec.prologue::prepare_transaction
> 
>   exec.note_setup::prepare_note
>   call.note1::main
>   
>   exec.note_setup::prepare_note
>   call.note2::main
> 
>   exec.epilogue::finalize_transaction
> end
> ```

yep, I left out the other components for brevity. 

> Second, I'm not sure we actually ever need to deal with textual representation of the above. The way I'm imagining we'd build a transaction program is not by writing the above MASM and compiling it, but by combining blocks manually. Specifically:
> 
> * We'd have a pre-build `CodeBlock`'s for prologue, epilogue, and note setup script (together with their associated callsets).

Having pre-compiles for this makes a lot of sense, I like this idea!

One way we could support this in the `Assembler` is to introduce a `pre_compiles` component:

```rust
pub struct Assembler<'a> {
    kernel: Kernel,
    module_provider: ModuleProvider,
    proc_cache: RefCell<ProcedureCache>,
    in_debug_mode: bool,
    pre_compiles: Option<&'a ProcedureCache>,
}
```

> * For each transaction, we'd build `CodeBlock`'s for each note using `compile_in_context()`.

I'm guessing we would then need to also return the `CodeBlockTable` for each note such we have access to the call target procedures?  In essence we would return a `Program` struct?  If we only return the `CallSet` then I'm not sure how the client will have access to the associated procedures?

> * And then we'd "manually" join these blocks together using constructors of `new_join()` and `new_call()` constructors of `CodeBlock` (we'd also need to merge their respective callsets).
> 
> To perform the above procedure, we could have a component in `miden-base` (or somewhere else) - maybe something like `TxCompiler` which would internally maintain pre-built `CodeBlock`'s for prologue, epilogue etc. as well as an instance of the assembler.

It seems to me that the `TxCompiler` would be replicating a lot of logic of the `Assembler`.  On the surface I think it would be easier to extend the `Assembler` to support this compilation pattern then to create a `TxCompiler` object.    

> Btw, with the introduction of `pcall` much of the above complexity would go away as we'd have a single program for transaction execution which could look something like:
> 
> ```
> use.miden::sat::internal::prologue
> use.miden::sat::internal::epilogue
> use.miden::sat::internal::note_setup
> 
> begin
>   exec.prologue::prepare_transaction
> 
>   while.true
>     exec.note_setup::prepare_note
>     pcall
>   end
> 
>   exec.epilogue::finalize_transaction
> end
> ```

Agreed!  `pcall` will make things easier, we will just need to specify call targets such that the `CodeBlockTable` can be constructed appropriately.

bobbinth reviewed (commented):
Thank you! I left a couple of comments inline. Overall, there are two open questions in my mind still:

1. Do we actually need to load account procedures into the assembler cache by name? It seems like it could lead to collisions. One example of such a collision I listed in the comments (and I think we could probably find a way how to fix that), but another example would be if we want to load multiple accounts into the assembler and they might have procedures with similar names but different implementations. One way to handle this would be to first address [#904](https://github.com/0xPolygonMiden/miden-vm/issues/904) and then come back to implement this using the `call.0x1234567` syntax.
2. How do we build transaction programs? My initial thinking was that we build them manually by joining compiled note script `CodeBlock`'s and `CodeBlock`'s for prologue, epilogue etc. Under this approach, we won't need to modify the assembler as much (as we won't need to import note scripts via `use.context::note1`) and it would make transition to `pcall` a bit simpler. Though, the code for manual joining would probably be a bit more complex than using the assembler.

frisitano commented on assembly/src/tests.rs:
```diff
@@ -105,6 +108,71 @@ fn span_and_simple_if() {
+        LibraryPath::new("context::note1").unwrap(),
+        procedure_name.clone(),
+    );
+    let note_1 =
+        ProgramAst::parse("use.context::account begin exec.account::account_method_1 end").unwrap();
+    let _note_1_root = assembler.compile_in_context(note_1, &mut assembler_context).unwrap();
```
> Also, I wonder if we should use a different path which can ensure that we don't run into collisions. In this specific case, `account` namespace will probably conflict with `miden::sat::account`. Eventually, these calls should be replaced by `call.0x1234567` format - but I'm actually unsure what to do in the meantime.

I think this illuminates a limitation of the current implementation.  Methods in `context::account` and `miden::sat::account` would not conflict in the `ProcedureCache` as we use the [fully qualified module path](https://github.com/0xPolygonMiden/miden-vm/blob/next/assembly/src/assembler/context.rs#L363-L379) when constructing the `ProecedureId`.  However, the way in which we define our call targets uses only the [module and procedure name](https://github.com/0xPolygonMiden/miden-vm/blob/next/assembly/src/parsers/context.rs#L154-L185) - we then use the module to fetch the full qualified path of the import.  Maybe it makes sense to use fully qualified paths for call targets to disambiguate such conflicts?  E.g. I should be able to use both `context::account` and `miden::sat::account` in the same module:

```
use.miden::sat::account
use.context::account

begin
     call.miden::sat::account::nice_procedure
     call.context::account::evil_procedure
end
```

frisitano commented:
> 1. Do we actually need to load account procedures into the assembler cache by name? It seems like it could lead to collisions. One example of such a collision I listed in the comments (and I think we could probably find a way how to fix that), but another example would be if we want to load multiple accounts into the assembler and they might have procedures with similar names but different implementations. One way to handle this would be to first address [#904](https://github.com/0xPolygonMiden/miden-vm/issues/904) and then come back to implement this using the `call.0x1234567` syntax.

As noted in the reply to your comment, provided the fully qualified module path is unique then I don't think this will lead to collisions in the `ProecedureCache` (however it will do in the parser).  If we needed to load two accounts then we could specify non-conflicting module paths appropriately - potentially using `account_id` as part of the path (e.g. `context::account::account_id`).  I'm not convinced about the usage of using MAST roots for call targets in this context - it seems to me that it reduces readability.  Is there a hard requirement for their usage?

> 2. How do we build transaction programs? My initial thinking was that we build them manually by joining compiled note script `CodeBlock`'s and `CodeBlock`'s for prologue, epilogue etc. Under this approach, we won't need to modify the assembler as much (as we won't need to import note scripts via `use.context::note1`) and it would make transition to `pcall` a bit simpler. Though, the code for manual joining would probably be a bit more complex than using the assembler.

I think the first question we should answer here is should the `Assembler` be responsible for the construction of the transaction program.  In my opinion, the introduction of a `TxCompiler` indicates some limitation of the `Assembler`. The transaction program is a `Program` like any other and as such we should be able to compile it using the `Assembler`.  On the other hand I guess it could be argued that the compilation pattern for the transaction program is unique enough that it is out of scope for the `Assembler` to support.  However, I tend to lean on the side of the former opinion - the `Assembler` should be flexible enough to support this.

bobbinth commented:
> I'm not convinced about the usage of using MAST roots for call targets in this context - it seems to me that it reduces readability. Is there a hard requirement for their usage?

I don't think it is a hard requirement - but I do think it is a better overall approach. Here is how I'm thinking about it:

In my mind, procedures which an account can expose fall into two categories:
1. Standardized procedures - e.g., `miden::abi::basic_wallet`.
2. Non-standard, account-specific procedures (similar to the ones you have in the tests here).

Calling procedures from the first category should probably be done using the approach we have now - i.e., `call.basic_wallet::receive_assets`. This has nice advantage of not tying the call to a specific account - i.e., this call can be done against any account which exposes `miden::abi::basic_wallet::recieve_assets` procedure (the resolution will be done via MAST root, but procedure ID reference is also unambiguous).

For the second category, things are a bit more complicated.

First, if we use something like `context::account` we may run into collision issues at the parser level. Also, in the above `context` now has a special meaning (i.e., nobody should create a library with `context` as a root namespace) and having it as such feels a bit brittle. So far, for special names, I've tried to use something that couldn't be declared by the user (e.g., `#sys` or `#exec`). We could do the same for `context` - i.e., use `#context` - but this would require quite a few updates to the parsing logic (and wouldn't solve parser collisions mentioned above).

Another issue is that if we have a user wallet which manages several accounts (or a block producer which needs to execute public transactions against many accounts), we'd probably want to re-use the same instance of the assembler between them (to avoid re-compiling common procedures). With `context::account` we cannot do that because the `account` part would actually be different for different accounts. A way to avoid it is to use account ID somewhere in the path. For example, something like this:
```
use.context::0x1234567 # this is the account ID

begin
  ...
  call.0x1234567::procedure_1
  ...
end
```
This fully disambiguates procedure IDs, but has a couple of issues of its own:
1. Whoever crafts the note needs to know the account ID. This may be both difficult and undesirable (for privacy reasons).
2. We cannot execute such notes against multiple accounts who share the same (though non-standard) interface.

Overall, for calling non-standard procedures, we probably want to have the following properties:
- Ability to call a procedure without having to specify an account ID.
- Ability to call a procedure against multiple accounts which expose the same non-standard interface.
- Ability to have multiple accounts loaded into the assembler cache while keeping references unambiguous.

I think that `call.0x1234567` (where `0x1234567` is the MAST root of the procedure) approach achieves these goals. We basically say that we can execute this note against any account which exposes a procedure with the specified MAST root. Another nice benefit is that it binds the call very explicitly and avoids potential issues which we could run into for accounts with updatable code. For example, say our note script is written as:
```
use.context::0x1234567 # this is the account ID

begin
  ...
  call.0x1234567::procedure_1
  ...
end
```
If account `0x1234567` has updatable code, the implementation of `procedure_1` could be changed to mean something completely different from the original procedure. Then, if we are not careful, we may compile our note against the new implementation without realizing that it is no longer what we want.

For standardized procedures this should be much less of an issue because, hopefully, these either don't change or change very infrequently with enough lead time for users to adjust (and if someone wants, they can also call standardized procedures by their hash rather than procedure name).

I do agree that readability suffers somewhat - though, I hope that this can be addressed via tooling / higher-level languages.

frisitano commented:
> Calling procedures from the first category should probably be done using the approach we have now - i.e., `call.basic_wallet::receive_assets`. This has nice advantage of not tying the call to a specific account - i.e., this call can be done against any account which exposes `miden::abi::basic_wallet::recieve_assets` procedure (the resolution will be done via MAST root, but procedure ID reference is also unambiguous).

Wouldn't we want some compile time guarantees that the `miden::abi::basic_wallet::recieve_assets` procedure is actually exposed by the account interface?  A user may accidentally use some procedure which is not actually part of the target account interface.

> For the second category, things are a bit more complicated.
> 
> First, if we use something like `context::account` we may run into collision issues at the parser level. Also, in the above `context` now has a special meaning (i.e., nobody should create a library with `context` as a root namespace) and having it as such feels a bit brittle. 

Agreed, I do think this is a bit brittle.  Having said that maybe we could leverage the [`add_module`](https://github.com/0xPolygonMiden/miden-vm/blob/next/assembly/src/assembler/module_provider.rs#L33) method of the `ModuleProvider` which checks for conflicts. 

>So far, for special names, I've tried to use something that couldn't be declared by the user (e.g., `#sys` or `#exec`). We could do the same for `context` - i.e., use `#context` - but this would require quite a few updates to the parsing logic (and wouldn't solve parser collisions mentioned above).

I think there is a bit of nuance to this.  `#sys` and `#exec` are special in the context of the virtual machine.  `context` is special in the context of the application (the miden rollup).  I don't think it makes sense to push application specific requirements into the `Assembler` the `add_module` solution referenced above would probably be fine.

> Another issue is that if we have a user wallet which manages several accounts (or a block producer which needs to execute public transactions against many accounts), we'd probably want to re-use the same instance of the assembler between them (to avoid re-compiling common procedures). With `context::account` we cannot do that because the `account` part would actually be different for different accounts. A way to avoid it is to use account ID somewhere in the path. For example, something like this:
> 
> ```
> use.context::0x1234567 # this is the account ID
> 
> begin
>   ...
>   call.0x1234567::procedure_1
>   ...
> end
> ```
> 
> This fully disambiguates procedure IDs, but has a couple of issues of its own:
> 
> 1. Whoever crafts the note needs to know the account ID. This may be both difficult and undesirable (for privacy reasons).

For the purpose of privacy I was of the understanding that it would be the receiver that would construct the note script and provide the `recipient` to the sender.  The assumption would then be that the receiver knows the account id. 

> 2. We cannot execute such notes against multiple accounts who share the same (though non-standard) interface.
> 
> Overall, for calling non-standard procedures, we probably want to have the following properties:
> 
> * Ability to call a procedure without having to specify an account ID.
> * Ability to call a procedure against multiple accounts which expose the same non-standard interface.
> * Ability to have multiple accounts loaded into the assembler cache while keeping references unambiguous.
 
One point around ambiguity here is we do not know which procedures are associated with which accounts, we will just know that at least one of the accuonts loaded in the assembler cache has this procedure.

> I think that `call.0x1234567` (where `0x1234567` is the MAST root of the procedure) approach achieves these goals. We basically say that we can execute this note against any account which exposes a procedure with the specified MAST root. Another nice benefit is that it binds the call very explicitly and avoids potential issues which we could run into for accounts with updatable code. For example, say our note script is written as:
> 
> ```
> use.context::0x1234567 # this is the account ID
> 
> begin
>   ...
>   call.0x1234567::procedure_1
>   ...
> end
> ```
> 
> If account `0x1234567` has updatable code, the implementation of `procedure_1` could be changed to mean something completely different from the original procedure. Then, if we are not careful, we may compile our note against the new implementation without realizing that it is no longer what we want.

Agreed but I would say this is an edge case.  The base case is worse for MAST calls as we do not know the context / namespace from which the MAST procedure originates - i.e. as mentioned before we do not know the MAST procedure is part of the target account interface.

I see pros and cons of both approaches here.  My conviction around either approach isn't particularly high so I'm happy to take your lead. I will back out changes related to tx program compilation.

bobbinth commented:
> In my opinion, the introduction of a `TxCompiler` indicates some limitation of the `Assembler`.

In my mind, `TxCompiler` is a bit more general. Specifically, I'm imagining a component which takes an account (or maybe just an account ID) and a set of notes, and outputs a [CompiledTransaction](https://github.com/0xPolygonMiden/miden-base/blob/main/objects/src/transaction/compiled_tx.rs) object. Internally, it maintains an instance of the assembler to perform compilation of note scripts etc. but it may be responsible for other things (e.g., loading account modules into the assembler). For example, part of the interface could look something like this:

```Rust
impl TxCompiler {
    pub fn load_account(&mut self, account_code: ModuleAst) -> AccountCode {
        todo!()
    }

    pub fn compile_transaction(
        &self,
        account_id: AccountId,
        notes: Vec<Note>,
        tx_script: Option<ProgramAst>,
    ) -> CompiledTransaction {
        todo!()
    }
}
```
If it can push more work onto the assembler, I think that's fine - but it probably still needs to do a few things to orchestrate the compilation process.

bobbinth commented:
> Wouldn't we want some compile time guarantees that the `miden::abi::basic_wallet::recieve_assets` procedure is actually exposed by the account interface? A user may accidentally use some procedure which is not actually part of the target account interface.

Yes, we would - but I'm not sure this could be guaranteed by the assembler (or via MASM code in general). I think we'll need to have something else that is aware of the state of the blockchain to perform these checks. (maybe it could be a part of the `TxCompiler` mentioned in the comment above).

> For the purpose of privacy I was of the understanding that it would be the receiver that would construct the note script and provide the `recipient` to the sender. The assumption would then be that the receiver knows the account id.

This is true. I was thinking more about a use case with multiple potential recipients - but that's a separate concern.

> One point around ambiguity here is we do not know which procedures are associated with which accounts, we will just know that at least one of the accuonts loaded in the assembler cache has this procedure.

Yeah, that's an issue - but I'm not sure how to get around this as in some cases we probably want to send a note to any account which can execute it (e.g., for a swap). So, we can't specify account ID upfront.

> I see pros and cons of both approaches here. My conviction around either approach isn't particularly high so I'm happy to take your lead. I will back out changes related to tx program compilation.

Yeah - I also see pros and cons. I do think MAST-based calls are better than using `context::account` imports, but they are not ideal either. Maybe there is a better solution which accommodates all use cases we are thinking about.

bobbinth commented:
> I do think MAST-based calls are better than using `context::account` imports

And just to state the reason concisely: I think we should be able to send notes to multiple recipients (i.e., any account which implements a given interface, which may or may not have been standardized) and MAST-based calls support this while I'm not sure how it could work with `context::account` import approach.

bobbinth commented:
> One point around ambiguity here is we do not know which procedures are associated with which accounts, we will just know that at least one of the accuonts loaded in the assembler cache has this procedure.

To come back to this again: I think this is a real issue and we should avoid this for sure. I'll need to think a bit more about this.

bobbinth commented:
When compiling notes (e.g., note `x` against account `a`), we'd like to achieve the following goals:

1. We may want to get compile-time guarantees that `x` can be executed against `a` and only against `a`.
2. If the above is not possible, we'd like to get guarantees that `x` can be executed against `a` (and maybe against other accounts).
3. If we cannot do the above, what can we do?

Let's look into these one-by-one.

Given the current structure of MASM, it is pretty clear that achieving the first point is impossible. If `x` calls `a.foo`, there is nothing preventing account `b` from also implementing `foo`. The only way we can ensure that `x` can be executed only against `a` is to check some properties of `a` dynamically. For example, something like this:
```
use.miden::sta::account
use.context.0x1234567 # this is the account ID

begin
  exec.account::get_id
  assert_eq
  call.0x1234567::foo
end
```
The above assumes that the first note input is set to account ID of `a` - so, what we do here is compare ID of the current account to the ID provided via note inputs, and if they match, we call `a.foo`. Without knowing what note inputs are and executing the code, we won't know whether the note can be executed against `a` and only against `a`.

Moving on to the second point, I would argue that this is also impossible to do at compile time. Consider a note with the following script (there are two accounts both implementing `foo`):
```
use.context.0x1234567
use.context.0x7654321

begin
  push.1
  if.true
    call.0x1234567::foo
  else
    call.0x7654321::foo    
  end
end
```
This note would compile against both accounts, but clearly, it can be executed only against one. The above is just a simple illustrative example (which could be caught by dead code elimination), but real notes could be way more complex with branching conditions coming from note inputs.

There is one other thing that I find problematic with the above code: it gives a perception that `foo` does the same thing when called as `0x1234567::foo` and as `0x7654321::foo` - though, in fact these could be very different procedures. In fact, if we are compiling this note in the context of `0x1234567`, we may not even know what `0x7654321::foo` is (i.e., if `0x7654321` is a private account).

So, if we can't guarantee at compile time whether `x` can be executed against `a` - what can we do?

For one, we should be able to tell if `x` cannot be executed against `a`. Let's consider this script for example (here I'm using MAST root call notation):
```
begin
  call.0x123456789
end
```
At compile time, it should be pretty easy to tell if this note can be executed against `a`. To do this, all we need to do is check if `a` exposes a procedure with MAST root `0x123456789`.

This also means that we should be able to perform a similar check against:
```
use.miden::abi::basic_wallet

begin
  call.basic_wallet::recieve_assets
end
```
This works because we can statically resolve `miden::abi::basic_wallet::recieve_assets` to its MAST root.

The situation is a bit more tricky if we have something like this:
```
begin
  <some code here>
  if.true
    call.0x123456789 # is expected to be called on account 0x123
  else
    call.0x987654321 # is expected to be called on account 0x456
  end
end
```
Without extra information, the best thing we can do is to check whether `a` exposes either procedure `0x123456789` or procedure `0x987654321`, and if it does, we can assume that it may be executed. This is because we don't know which branch will be taken at compile time. There could be more complex situations but as long as all procedures in one branch can be executed against `a`, we assume that `x` may be executed against `a` and this can be checked using fairly simple static analysis.

This static analysis cannot be performed by the assembler (because the assembler doesn't know these rules) - but it can be performed by something like a `TxCompiler`. One way to do it would be this:
1. Compile the program with the assembler.
2. Traverse the MAST in DFS order.
3. Use `CALL` nodes to build a set of calls made in the same branch.
4. Check if all targets of these calls are present in target account's interface.

We could also improve on the above by providing the extra information in the program itself. For this, we should probably use new syntax. Maybe something like this:
```
begin
  <some code here>
  if.true
    call.0x123456789@0x123
  else
    call.0x987654321@0x456
  end
end
```
During compilation to MAST, the assembler could add these attributes to `CALL` nodes, and then `TxCompiler` would interpret them accordingly. Though it is worth noting again that with this type of static analysis we can can't give true guarantees that we'll be able to execute a note against `a` - but we may be able to catch more instances of when we wouldn't be able to do so.

bobbinth commented:
Assuming my last comment was not too wrong, I think the way to proceed on this is:

1. Keep this PR open for now.
2. Implement simple MAST-based `call` instruction (without attributes). This will require [#904](https://github.com/0xPolygonMiden/miden-vm/issues/904).
3. Merge the result into this PR and update accordingly.
4. Merge this PR into `next`.
5. In the future, add attribute syntax to MAST-based call instruction.

bobbinth commented on assembly/src/assembler/mod.rs:
```diff
@@ -148,15 +148,33 @@ impl Assembler {
         Ok(Program::with_kernel(program_root, self.kernel.clone(), cb_table))
     }
 
+    /// Compiles the provided [ProgramAst] into a [CodeBlock].
+    pub fn compile_program_ast(&self, program: ProgramAst) -> Result<CodeBlock, AssemblyError> {
+        let (local_procs, body) = program.into_parts();
+
+        let mut context = AssemblyContext::new(AssemblyContextType::Program);
+        // compile all local procedures; this will add the procedures to the specified context
+        for proc_ast in local_procs.iter() {
+            if proc_ast.is_export {
+                return Err(AssemblyError::exported_proc_in_program(&proc_ast.name));
+            }
+            self.compile_procedure(proc_ast, &mut context)?;
+        }
+
+        // compile the program body
+        let program_root = self.compile_body(body.iter(), &mut context, None)?;
+
+        Ok(program_root)
+    }
```
The intent of this is to be used for compiling note scripts, right? I'm thinking maybe we should change it back to a version of `compile_in_context` we had before. It doesn't need to be very different from what you have now - e.g., it could look something like this:
```Rust
pub fn compile_in_context(
    &self,
    program: ProgramAst,
    context: &mut AssemblyContext,
) -> Result<CodeBlock, AssemblyError> {
    // TODO: make sure the context is for compiling programs

    let (local_procs, body) = program.into_parts();

    // compile all local procedures; this will add the procedures to the specified context
    for proc_ast in local_procs.iter() {
        if proc_ast.is_export {
            return Err(AssemblyError::exported_proc_in_program(&proc_ast.name));
        }
        self.compile_procedure(proc_ast, context)?;
    }

    // compile the program body
    let program_root = self.compile_body(body.iter(), context, None)?;

    Ok(program_root)
}
```

Then, we can change the `compile()` method to look something like this:
```Rust
  pub fn compile<S>(&self, source: S) -> Result<Program, AssemblyError>
  where
      S: AsRef<str>,
  {
      // parse the program into an AST
      let source = source.as_ref();
      let program = ProgramAst::parse(source)?;

      // compile the program
      let mut context = AssemblyContext::new(AssemblyContextType::Program);
      let program_root = self.compile_in_context(program, &mut context)?;

      // convert the context into a call block table for the program
      let cb_table = context.into_cb_table(&self.proc_cache.borrow());

      // build and return the program
      Ok(Program::with_kernel(program_root, self.kernel.clone(), cb_table))
  }
```

frisitano replied:
I like the idea of calling `compile_in_context(..)` from `compile(..)` but I'm not sure if the caller should be responsible for managing the context.  Maybe instead the `compile_program_ast(..)` should return both `program_root` and `cb_table`.  This would become:

```rust
  pub fn compile<S>(&self, source: S) -> Result<Program, AssemblyError>
  where
      S: AsRef<str>,
  {
      // parse the program into an AST
      let source = source.as_ref();
      let program = ProgramAst::parse(source)?;

      // compile the program
      let mut context = AssemblyContext::new(AssemblyContextType::Program);
      let (program_root, cb_table) = self.compile_program_ast(program, &mut context)?;

      // build and return the program
      Ok(Program::with_kernel(program_root, self.kernel.clone(), cb_table))
  }
```

Are there any benefits to having the caller manage the context?

bobbinth replied:
I'm thinking of how we'd use it down the road to construct transaction programs. Whether we do it by manually combining `CodeBlock`'s or via `pcall` instruction in the future, we'd still need to build up a callset of all invoked procedures. `AssemblyContext` already keeps track of this - so, I think it would a natural way to build a unified callset across all notes - though, we'd have to make a few updates before we can use it like this.

frisitano replied:
This makes sense.  I think we will need to make some changes as the `into_cb_table(..)` currently consumes the context but this shouldn't be a problem.

bobbinth reviewed (changes requested):
Looks good! Thank you! I left a couple of comments inline - let me know what you think.

bobbinth commented:
> @bobbinth if you could support with the failing test that would be appreciated.

I pushed a commit which fixes the compilation error. The test which uses this code is currently ignored (as it is failing and need to be updated to include latest changes from Winterfell). This is something @Al-Kindi-0 will address in future PRs.

bobbinth commented on assembly/src/tests.rs:
```diff
@@ -105,6 +108,50 @@ fn span_and_simple_if() {
+        .unwrap();
+
+    // compile note 1 program
+    let note_1 =
+        ProgramAst::parse("use.context::account begin call.account::account_method_1 end").unwrap();
+    let _note_1_root = assembler
+        .compile_in_context(note_1, &mut super::AssemblyContext::new(AssemblyContextType::Program))
+        .unwrap();
+
+    // compile note 2 program
+    let note_2 =
+        ProgramAst::parse("use.context::account begin call.account::account_method_2 end").unwrap();
+    let _note_2_root = assembler
+        .compile_in_context(note_2, &mut super::AssemblyContext::new(AssemblyContextType::Program))
+        .unwrap();
```
Ideally, we should compile both notes in the same context - but I think it is fine for now.

bobbinth reviewed (approved):
Thank you! Looks good! Once the next version of Winterfell is released, I'll make a commit here to use the new version in dependencies and will merge after that.
